### PR TITLE
Require device trust for initial device registration endpoints

### PR DESCRIPTION
**Pull request by Joerger**
Changelog: When device trust is required and MFA is optional, users will need to add their first MFA device from a trusted device.

Note: This this will make it impossible for a user to add their first MFA device from the WebUI, as device trust is not yet supported in the WebUI. They can instead add it with `tsh`. It may also be worth adding device management to Teleport Connect. A fix for device trust in the WebUI may be developed soon - https://github.com/gravitational/teleport.e/issues/3236.

### CODE REVIEW

**Review by avatus — COMMENTED** [review]
Code looks fine but I'm not sure we want to disable first-time registration on web until another option is in place. Although, that isn't up to me I suppose! I don't feel comfortable approving as my context here is pretty light.

**Review by codingllama — COMMENTED** [review]
Looks good. No product commentary on my part either.

> Code looks fine but I'm not sure we want to disable first-time registration on web until another option is in place

@avatus note that, as this is written, enforcement is based on the global device trust mode, so that does alleviate the "breakage" somehow.

File: lib/auth/auth_with_roles.go [hunk 1]
```diff
@@ -6318,6 +6318,13 @@ func (a *ServerWithRoles) CreateAuthenticateChallenge(ctx context.Context, req *
 
 // CreatePrivilegeToken is implemented by AuthService.CreatePrivilegeToken.
 func (a *ServerWithRoles) CreatePrivilegeToken(ctx context.Context, req *proto.CreatePrivilegeTokenRequest) (*types.UserTokenV3, error) {
+	// Device trust: authorize device before issuing a privileged token without an MFA response.
+	if mfaResp := req.GetExistingMFAResponse(); mfaResp == nil || (mfaResp.GetTOTP() == nil && mfaResp.GetWebauthn() == nil) {
```
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/auth/auth_with_roles.go:6322]: Wouldn't this give a null pointer exception if `mfaResp == nil`? Or do the proto getter methods actually protect from that?

[the remark this answers is not in the document]
Comment (codingllama) [reply lib/auth/auth_with_roles.go:6322]: Proto getters are safe against nil.

File: lib/auth/auth_with_roles.go [hunk 2]
```diff
@@ -6337,6 +6350,15 @@ func (a *ServerWithRoles) CreateRegisterChallenge(ctx context.Context, req *prot
 	return a.authServer.CreateRegisterChallenge(ctx, req)
 }
 
+func (a *ServerWithRoles) authorizeDevice(ctx context.Context) error {
+	authPref, err := a.authServer.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = dtauthz.VerifyTLSUser(authPref.GetDeviceTrust(), a.context.Identity.GetIdentity())
```
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/auth/auth_with_roles.go:6358]: On second thought we probably should actually exercise the role device mode, will update.

File: lib/auth/auth_with_roles.go [hunk 3]
```diff
@@ -6318,6 +6318,13 @@ func (a *ServerWithRoles) CreateAuthenticateChallenge(ctx context.Context, req *
 
 // CreatePrivilegeToken is implemented by AuthService.CreatePrivilegeToken.
 func (a *ServerWithRoles) CreatePrivilegeToken(ctx context.Context, req *proto.CreatePrivilegeTokenRequest) (*types.UserTokenV3, error) {
+	// Device trust: authorize device before issuing a privileged token without an MFA response.
```
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/auth/auth_with_roles.go:6321]: `CreatePrivilegeToken` is used to create a token to then create a register challenge without the normal MFA check, which includes the check to see if the user already has no MFA device yet. I could refactor it a bit to only impact `CreateRegisterChallenge`, let me see if it's not too much work.

File: lib/auth/grpcserver_test.go [hunk 4]
```diff
@@ -976,6 +977,123 @@ func TestGenerateUserCerts_deviceAuthz(t *testing.T) {
...
+		TestBuildType: modules.BuildEnterprise, // required for Device Trust.
+	})
+
+	testServer := newTestTLSServer(t)
+
+	ctx := context.Background()
+	authServer := testServer.Auth()
+
+	// Create a user for testing.
+	user, role, err := CreateUserAndRole(testServer.Auth(), "llama", []string{"llama"}, nil)
```
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/auth/grpcserver_test.go:992]: Hope you don't mind me using your namesake :llama:

File: lib/auth/auth_with_roles.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/auth/auth_with_roles.go:6358]: I changed my mind, this device trust requirement is only useful in cases where it is enforced globally and not by role. Thanks for ensuring we're on the same page, it makes more sense now.

File: lib/auth/auth_with_roles.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Joerger) [reply lib/auth/auth_with_roles.go:6321]: So as it turns out, we do need to do the device trust check on `CreatePrivilegeToken`. This endpoint uses the user's credentials to generate a privileged Token either. An MFA check is required if the user has a registered device, but otherwise a `privilege_exception` token is created without the MFA check.

The Proxy then uses the user-created token to create a register challenge on behalf of the user. Enforcing device trust at this step won't work, so instead we do it when the `privilege_exception` token is created.

I had attempted some refactors to consolidate logic, but it all fell through once I realized this limitation.

**Review by avatus — APPROVED** [review]

File: lib/auth/auth_with_roles.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/auth/auth_with_roles.go:6321]: Could we leave a comment here saying that we do this check because of privilege tokens? (Ie, a summary of what you just explained?)

File: lib/auth/auth.go [hunk 5]
```diff
@@ -3112,9 +3112,6 @@ func (a *Server) CreateRegisterChallenge(ctx context.Context, req *proto.CreateR
 		if err != nil {
 			return nil, trace.Wrap(err)
 		}
-
-	default:
-		return nil, trace.BadParameter("either a token or an MFA response are required")
```
Comment (codingllama) [inline lib/auth/auth.go:3117]: Is this change necessary? Just curious.

**Review by codingllama — APPROVED** [review]

File: lib/auth/auth.go (hunk 5 above)
Comment (Joerger) [reply lib/auth/auth.go:3117]: Only to make this change work - https://github.com/gravitational/teleport/pull/38451#discussion_r1499391831

Should I revert?

Comment (codingllama) [reply lib/auth/auth.go:3117]: Nope, it's fine.
